Add Variables

kinseyda opened the pull request:
I added functionality to allow for user defined variables, accessible by prefixing with a $
This includes
- Recalling the previous evaluated answer using `$ans`.
- Assigning any value (including the result of a expression) to a variable with an arbitrary name given by the user.
        -  e.g. `>$num = 5` and recalling with `2 + $num + 3`
- Reassigning the value of a variable using the same syntax.

Commit:
add stuff

Commit:
Merge branch 'master' of https://github.com/George-lewis/rustcalc into master

Commit:
Add struct for variables

Commit:
Revert "Add struct for variables"

This reverts commit 5bb06985080689ad02bba79f00d6d414184c11b4.

Commit:
Add variable recognition/tokenization

Commit:
Add variable list command

Commit:
Make list command appear in history

Commit:
Change starting variable to ans

Name only at the moment, no functionality

Commit:
Add variable assignment

Commit:
Move error handling to separate function

Commit:
Allow variables to be assigned with expressions

Commit:
Add ans functionality

Commit:
Add variable reassigning

Commit:
Reduce visible decimal places in variable list command

Commit:
Add assignment confirmation

Commit:
Add error handling

Commit:
Add $ to ideal representation and user input requirement

Commit:
Remove unused import

Commit:
Add test cases for variable tokenization

Commit:
Remove unused variable naming error

Commit:
Add more comments

George-lewis committed:
Merge branch 'variables' into Add-variables

George-lewis committed:
Merge branch 'master' into Add-variables

Commit:
Fix formatting

George-lewis commented on src/lib/errors.rs:
```diff
@@ -6,4 +6,6 @@ pub enum Error {
     Operand(OperatorType),
     EmptyStack,
     MismatchingParens,
+    Assignment,
```
The `/lib` directory should be treated as an entirely separate logical unit from the CLI code (when in CLI code think of it as an external dep / black box). The CLI just consumes the library. Thus, this error enum is only for library errors.

George-lewis commented on src/lib/mod.rs:
```diff
@@ -7,15 +7,17 @@ mod rpn;
 mod tokenize;
 pub mod tokens;
 pub mod utils;
+pub(crate) mod variables;
```
```suggestion
pub mod variables;
```

George-lewis commented on src/lib/tokenize.rs:
```diff
@@ -46,6 +45,19 @@ pub fn tokenize(string: &str) -> Result<Vec<Token>, Error> {
         // Slice the input from the index until the end
         let slice = utils::slice(string, idx, -0);
 
+        let kind;
```
Not idiomatic. In Rust you can assign to expressions, like:

```rust
let kind = if condition {
  a
} else {
  b
}
```

kinseyda replied:
Fixed as of 6680a94a7db7e5138fa0ee1ebec27250b4c9a426

George-lewis commented on src/lib/tokenize.rs:
```diff
@@ -46,6 +45,19 @@ pub fn tokenize(string: &str) -> Result<Vec<Token>, Error> {
 
+        let kind;
+
+        if c =='$' {
```
Refactor into `Variable::is` and then move it into `_type()`

kinseyda replied:
Fixed as mentioned, see 53aa5e7903f46d75699345d5f6c8336428b09bdd

George-lewis commented on src/lib/tokenize.rs:
```diff
@@ -66,12 +78,7 @@ pub fn tokenize(string: &str) -> Result<Vec<Token>, Error> {
             coeff = false;
         }
 
-        let kind = match _type(&slice) {
```
Why was this moved?

kinseyda replied:
The kind variable assignment has been moved back to right before it is needed as it was originally, see 53aa5e7903f46d75699345d5f6c8336428b09bdd

George-lewis commented on src/lib/tokenize.rs:
```diff
@@ -131,6 +138,18 @@ pub fn tokenize(string: &str) -> Result<Vec<Token>, Error> {
+                    Some(a) => a,
+                    None => return Err(Error::UnknownVariable(idx)),
+                };
+                idx += n+1; //+1 to account for '$'
```
Please insert a space between `//` and the start of a comment.

kinseyda replied:
Fixed as of 2b46055758992c2ff671cdfb1b9ebbbc810b0007

George-lewis commented on src/lib/tokens.rs:
```diff
@@ -64,6 +66,7 @@ impl Token {
                 ParenType::Right => ")".to_string(),
             },
             Self::Constant { kind } => Constant::by_type(*kind).repr[0].to_string(),
+            Self::Variable { inner } => "$".to_owned() + &inner.repr.to_string(),
```
```suggestion
            Self::Variable { inner } => format!("${}", inner.repr),
```

kinseyda replied:
Fixed as of 76048df83f3dc74449819f2133a17b18e04f614b

George-lewis commented on src/lib/variables.rs:
```diff
@@ -0,0 +1,33 @@
+#![allow(clippy::non_ascii_literal)]
```
Might be unnecessary. Try removing and running `cargo clippy`.

kinseyda replied:
You're right, fixed with 695aa48043b171c3e83792d39d662ac2bab65576

George-lewis commented on src/main.rs:
```diff
@@ -29,6 +30,15 @@ fn main() -> ! {
         editor.load_history(path).ok();
     }
 
+    let mut vars: Vec<Variable> = Vec::new();
```
1. Use the `vec!` macro
2. Use `String::from` for constant `String` values

```suggestion
    let mut vars = vec![
      Variable {
        repr: String::from("ans"),
        value: 0.0
      }
    ]
```

kinseyda replied:
Good call, much neater. Changed with 127dbe606ed8df144f0be1a1471a82e7dbdc8190

George-lewis commented on src/main.rs:
```diff
@@ -45,40 +55,74 @@ fn main() -> ! {
+            }
+            // Add the line to the history
+            editor.add_history_entry(input);
+            continue;
```
Probably better to do if/else. You could also factor out the `add_history_entry` line as a result.

kinseyda replied:
Resolved with 7974f77fc31db1279175c5e2d3ed34ab9f8fff3a, input handling now works via `if`/`else if`/`else `statements and not `continue`s

George-lewis commented on src/main.rs:
```diff
@@ -45,40 +55,74 @@ fn main() -> ! {
+            for v in vars.iter() {
+                println!(
+                    "[ {} => {} ]",
+                    ("$".to_owned() + &v.repr).green().bold(),
```
Use `format!`

kinseyda replied:
Fixed as of 76048df83f3dc74449819f2133a17b18e04f614b

George-lewis commented on src/main.rs:
```diff
@@ -45,40 +55,74 @@ fn main() -> ! {
-                        println!("Couldn't evaluate. Mismatched parens.");
-                    }
+        if input == "$" {
+            //Variable list command
```
```suggestion
            // Variable list command
```

kinseyda replied:
Fixed as of 2b46055758992c2ff671cdfb1b9ebbbc810b0007

George-lewis commented on src/main.rs:
```diff
@@ -45,40 +55,74 @@ fn main() -> ! {
+        }
+
+        if input.contains("=") {
+            //Variable assignment/reassignment
```
```suggestion
            // Variable assignment/reassignment
```

kinseyda replied:
Fixed as of 2b46055758992c2ff671cdfb1b9ebbbc810b0007

George-lewis commented on src/main.rs:
```diff
@@ -45,40 +55,74 @@ fn main() -> ! {
+
+        if input.contains("=") {
+            //Variable assignment/reassignment
+            let sides: Vec<&str> = input.split('=').collect();
```
Trim before collecting.

kinseyda replied:
The left hand side of the assignment is the only part that really needs to be trimmed (since variables need to start with $, and any white space after the variable name but before the = sign shouldn't be included in names), but I've reduced the calls to trim() to just 1 with be546457fbc7d9c02585f7e4bc34e5a73b9275a9.

George-lewis replied:
And we don't have to trim the right side because the tokenizer can deal with whitespace on its own? Hmmm, alright.

George-lewis commented on src/main.rs:
```diff
@@ -45,40 +55,74 @@ fn main() -> ! {
+            let (user_value, repr) = match doeval(&sides[1], &vars) {
+                Ok((a, b)) => (a, b),
+                Err(e) => {
+                    handle_errors(e, input.clone(), sides[0].len() + 1);
```
Why `+1`?

George-lewis commented on src/main.rs:
```diff
@@ -242,6 +360,23 @@ mod tests {
 
     #[test]
     fn test() {
+        let mut test_vars: Vec<Variable> = Vec::new();
```
Use `vec!` like at the top of `main()`.

kinseyda replied:
Resolved with 2fd49ed62a38ee138edbf263609df4c0e2aa2141

George-lewis commented on src/lib/variables.rs:
```diff
@@ -0,0 +1,33 @@
+impl Searchable for Variable {
+    fn search<'a, 'b>(&'a self, search: &'b str) -> Option<(&'a Self, usize)> {
+        // Case sensitive
+        let search_str = utils::slice(search, 1, -0); //to ignore $
```
We should probably be slicing in the outer layer (in the tokenizing code). `utils::slice` is specifically to handle unicode, however because we know the current character will be `$` in the relevant context you pay be able to get away with a cheaper slice: `&str[1..]`, which in theory should be as simple as passing a pointer...

kinseyda replied:
Resolved as stated with f895f1fc569766509614f612eb5b0d9e3ad7efd3

George-lewis commented on src/main.rs:
```diff
@@ -88,10 +132,79 @@ fn main() -> ! {
     }
 }
 
+fn handle_errors(e: Error, input: String, offset: usize) {
```
This feels kind of inverted, in `main()` you invoke this function directly in a lot of places. It would be better to encapsulate  most of the logic that's currently there into a new function, perhaps `fn handle_input(&str, ...?) -> Result<?>`. Feel free to define a new error enum specifically for the CLI half of the code. I can then imagine `main()` looking something like this:

```rust
fn handle_input(input: &str, vars: &mut Vec<Variable>) -> Result<String, CliError> {
  if input == "$" {
    formatted_vars_output
  } else if input.contains('=') {
    // Split it and do your stuff
    formatted_string
  } else {
    // Normal statement
    formatted_string
  }
}

fn handle_errors(error: CliError) {
  match error {
    ...
  }
}

fn main() {
  // Get input
  match handle_input(&input, &mut vars) {
    Ok(formatted) => println!(formatted),
    Err(error) => handle_error(error)
  }
}
```

But I'm not completely sure. I might need to think about this more.

kinseyda replied:
This has been fixed with 7974f77fc31db1279175c5e2d3ed34ab9f8fff3a and 104fb771e8228d22e69f3b2709d0d0f4d1ae61ca.
Mostly it is as you specified above though I've added no new errors, instead just bubbling up the existing ones. The code in the main loop is now a lot more logically laid out and should be easier to follow.

George-lewis commented on src/main.rs:
```diff
@@ -242,6 +360,23 @@ mod tests {
 
     #[test]
     fn test() {
```
Create a new test function for the cli / var specific stuff.

kinseyda replied:
Resolved with 2fd49ed62a38ee138edbf263609df4c0e2aa2141

George-lewis reviewed (changes requested):
Needs some more work. Most of it is general fixes like making things more idiomatic, or some reorganization, but there are a few major issues like the structure of `main.rs`.

kinseyda committed:
Update src/lib/mod.rs

Co-authored-by: George-lewis <33588728+George-lewis@users.noreply.github.com>

Commit:
Fix tokenize kind assignment

kinseyda committed:
Add comment

kinseyda committed:
Merge remote-tracking branch 'kinseyda_fork/Add-variables' into Add-variables

kinseyda committed:
Fix tokenize kind assignment expression syntax

kinseyda committed:
Fix comment formatting

kinseyda committed:
Fix tokenize comment formatting

kinseyda committed:
Move kind variable TokenType selection for variables to _type()

This makes variable recognition more consistent with the existing code for constants.

kinseyda committed:
Change string formatting pieces to use format!()

kinseyda committed:
Remove unneeded clippy change

kinseyda committed:
Simplify vars vector initialization

kinseyda committed:
Change where lines are added to history

This change means invalid inputs will now be remembered, allowing the user to change them quickly without retyping.

George-lewis commented on src/main.rs:
```diff
@@ -452,8 +633,9 @@ mod tests {
         ]
         .iter()
-        .for_each(|(a, b)| assert_eq!(doeval(a).unwrap_err(), *b));
+        .for_each(|(a, b)| assert_eq!(doeval(a, &Vec::new()).unwrap_err(), *b));
```
`&[]`

kinseyda replied:
Resolved with 2fd49ed62a38ee138edbf263609df4c0e2aa2141

kinseyda committed:
Refactor main loop

Main is now a lot cleaner,  with the input handling code having its own function which bubbles up results.

kinseyda committed:
Move variable assignment to separate function

George-lewis commented on src/main.rs:
```diff
@@ -100,61 +130,48 @@ fn main() -> ! {
 
-fn handle_errors(e: Error, input: String, offset: usize) {
-    // Offset is added so that highlighting can be added to expressions that come after an '=' during assignment
+fn handle_errors(e: Error, input: String) {
```
```suggestion
fn handle_errors(error: Error, input: &str) {
```

This function should probably take a `&str` instead, and as a personal preference I'd like to see a longer-than-one variable name here.

kinseyda replied:
Fixed with 16a1aaaa8cf7bb41804a6107c0c580ce5d3004b6

kinseyda committed:
Remove unnecessary trim() calls

kinseyda committed:
Add comment to trim() call

kinseyda committed:
Change handle_errors() signature

kinseyda committed:
Separate and simplify variable test cases

Moves test cases for variable tokenization to their own functions, and makes the variable list initialization more idiomatic.

kinseyda committed:
Change variable search to not ignore first character

This means that the $ prefix must be filtered off before calling next_variable()

kinseyda committed:
Fix formatting

George-lewis commented on src/main.rs:
```diff
@@ -56,42 +58,70 @@ fn main() -> ! {
+    if input == "$" {
+        // Variable list command
+        Ok(list_vars_command(&vars))
+    } else if input.contains("=") {
```
Can this be a char like line 103?

kinseyda replied:
Fixed with 24b79e5237619c3eed5608e18b574f2ee324ec76

George-lewis reviewed (changes requested):
Please add some docs on new functions, eventually I (or someone else?) will add docs to the rest.

Documentation in Rust is done with Markdown. You can see an example here, [the docs for `Vec::push`](https://doc.rust-lang.org/src/alloc/vec/mod.rs.html#1489). This can be rendered by IDEs or into websites like you can see [here](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.push).

George-lewis commented on src/lib/errors.rs:
```diff
@@ -6,4 +6,6 @@ pub enum Error {
     Operand(OperatorType),
     EmptyStack,
     MismatchingParens,
+    Assignment,
```
I hate to say it, but this feels like muddling CLI and backend concerns, given that assignment is something implemented purely in the interface code. Unfortunately we might have to nest errors here, something like:

```rust
enum CliError {
  Assignment,
  Library(Error)
}
```

which can be matched like:

```rust
match error {
  CliError::Assignment => ...,
  CliError::Library(inner) => {
    match inner {
      Error::EmptyStack => ...,
      ...
    }
  }
}
```

You can alleviate some of the pain this causes by [`impl From<Error> for CliError`](https://doc.rust-lang.org/std/convert/trait.From.html)

---

I wonder if there could be a better way, there's a lot of crates out there for graceful error handling, I'll have to look into it

kinseyda replied:
Resolved as stated, a new enum has been created to encompass errors in the library for code in the Cli, see f01deaaa794ccb84e35c1e8591a0e1e458a25552

kinseyda committed:
Change strings to chars where possible

kinseyda committed:
Add new CliError enum

The new CliError enum encompasses the library errors, separating them and errors that are unique to just the cli (currently this just means variable assignment errors).

George-lewis commented on src/main.rs:
```diff
@@ -149,55 +160,57 @@ fn handle_input(input: &str, vars: &mut Vec<Variable>) -> Result<String, Error>
+                } else {
+                    "".to_string()
+                };
+                println!(
```
It looks like this code that draws a red squiggle underneath the text is going to get reused a bit. Consider refactoring into a function.

kinseyda replied:
Fixed with 84c1762f24eb71a382465f427e23c0dd8e799487

George-lewis committed:
Merge branch 'variables' into Add-variables

kinseyda committed:
Move confirmation string generation out of handle_input()

kinseyda committed:
Merge remote-tracking branch 'kinseyda_fork/Add-variables' into Add-variables

kinseyda committed:
Move duplicated highlight-able error code to new function

kinseyda committed:
Move assignment code to separate function

This is mainly just to de-clutter the input handling function

kinseyda committed:
Fix error highlighting issue

Fixes the wrong character being highlighted when its on the right hand side of an assignment expression with added whitespace

kinseyda committed:
Simplify variable assign command call

kinseyda committed:
Fix formatting

George-lewis commented on src/main.rs:
```diff
@@ -45,50 +64,169 @@ fn main() -> ! {
+    let (user_value, repr) = result?;
+
+    // Get printable confirmation string
+    let conf_string = assign_var_conf_string(repr, &user_repr, user_value);
```
Inline this function

kinseyda replied:
I'm assuming you mean just put this line in the `Ok()` at the end? Correct me if I'm wrong, but I don't believe this is doable, since I need to drop the `repr` vector before I can mutate `vars` (the `assign_var()` call right after this line) because the lifetime of `repr` is the immutable borrow of  `vars` by `doeval()`

George-lewis replied:
I mean delete the function and insert the code here.

kinseyda replied:
Resolved with 674a8b7bd1af807a05c0fa85a8e71182d6f6b1be

kinseyda committed:
Add documentation to new functions, struct, and enums

George-lewis commented on src/lib/variables.rs:
```diff
@@ -18,12 +19,15 @@ impl Searchable for Variable {
 }
 
 impl Variable {
+    /// Searches for the first variable in `vars` that matches the representation given by `text`
```
+ text must start with the name of a variable (matching case) but can be arbitrarily long

kinseyda replied:
The documentation has been clarified with 5cb996f97f0584b65421db8089e7adb622a60222

kinseyda committed:
Remove string formatting function for var assignment

Moves this code back to assign_var_command(), but reduces it to only one assignment

kinseyda committed:
Clarify next_variable() documentation

George-lewis commented:
I found an issue. For some reason when the input string ends with a `$`, `make_highlighted_error` panics on line 191.

```
> 4$
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src\main.rs:191:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: process didn't exit successfully: `target\debug\rustcalc.exe` (exit code: 101)
```

kinseyda committed:
Change error highlighting for variables

This prevents a possible panic when the last character in a string is a $

kinseyda committed:
Change vars initialization to make $ans lazy

kinseyda committed:
Make vars list sorted by representation size

This means that the program will search for long variable names first, allowing for variable names to be sub-strings of each other

George-lewis commented on src/main.rs:
```diff
@@ -44,10 +44,7 @@ fn main() -> ! {
+    let mut vars = vec![];
 
     loop {
         #[allow(clippy::single_match_else)]
```
We can probably remove this now

George-lewis commented on src/main.rs:
```diff
@@ -167,7 +167,7 @@ fn handle_input(input: &str, vars: &mut Vec<Variable>) -> Result<String, CliErro
         let eval_string = format!("[ {} ] => {}", formatted, format!("{:.3}", x).blue());
 
-        vars[0].value = x; // Set ans to new value
+        assign_var(vars, x, "ans".to_string()); // Set ans to new value
```
`assign_var` should take a `&str` and create a `String` only if necessary

George-lewis committed:
make clippy happy, usr &str more

George-lewis committed:
lint

George-lewis committed:
use From impl

George-lewis committed:
rewrite assign_var

George-lewis committed:
adjust signature

George-lewis committed:
fix doc

George-lewis committed:
check for empty vars

George-lewis committed:
make clippy happy, remove unnecessary lifetimes

George-lewis committed:
make printing more consistent

George-lewis committed:
fix assign_var, lint

George-lewis committed:
simplify assign_var

George-lewis committed:
adjust assignment statement print

George-lewis commented:
Thank you for your help
